rest/admin_api.go

```diff
@@ -1568,13 +1568,55 @@ func (h *handler) updatePrincipal(name string, isUser bool) error {
 	if err != nil {
 		return err
 	} else if replaced {
+		// update event
+		if isUser {
+			base.Audit(h.ctx(), base.AuditIDUserUpdate, base.AuditFields{
+				"username": internalName,
+				"roles":    newInfo.ExplicitRoleNames.ToArray(),
+				"channels": getAuditEventAccess(h.db, isUser, newInfo),
+				"db":       h.db.Name,
+			})
+		} else {
+			base.Audit(h.ctx(), base.AuditIDRoleUpdate, base.AuditFields{
+				"role":           internalName,
+				"admin_channels": getAuditEventAccess(h.db, isUser, newInfo),
+				"db":             h.db.Name,
+			})
+		}
 		h.writeStatus(http.StatusOK, "OK")
 	} else {
+		// create event
+		if isUser {
+			base.Audit(h.ctx(), base.AuditIDUserCreate, base.AuditFields{
+				"username": internalName,
+				"roles":    newInfo.ExplicitRoleNames.ToArray(),
+				"channels": getAuditEventAccess(h.db, isUser, newInfo),
+				"db":       h.db.Name,
+			})
+		} else {
+			base.Audit(h.ctx(), base.AuditIDRoleCreate, base.AuditFields{
+				"role":           internalName,
+				"admin_channels": getAuditEventAccess(h.db, isUser, newInfo),
+				"db":             h.db.Name,
+			})
+		}
 		h.writeStatus(http.StatusCreated, "Created")
 	}
 	return nil
 }
 
+func getAuditEventAccess(db *db.Database, isUser bool, cnf auth.PrincipalConfig) map[string]map[string][]string {
+	auditEventAccess := make(map[string]map[string][]string)
+	collectionAccess := make(map[string][]string)
+	if db.OnlyDefaultCollection() {
+		collectionAccess[base.DefaultCollection] = cnf.ExplicitChannels.ToArray()
```

Comment:
This looks like it doesn't differentiate between a request that omits ExplicitChannels, and one that sets ExplicitChannels to an empty set (since ToArray() is going to return the same thing, a non-nil empty slice).